### ISSUE TITLE
All Properties Appear Changed, AuditExclude doesn't apply

### ISSUE THREAD

**Issue by mitchelsellers** [zzzprojects/EntityFramework-Plus#246]
Working with a .NET Core project, using EF Core. 

Audit works, however, all fields including key fields show as changed all the time.  

We have tried explicitly setting `AuditManager.DefaultConfiguration.IgnorePropertyUnchanged = true`  

We have even added the [AuditExclude] attribute to the object, and the requisite `AuditManager.DefaultConfiguration.ExcludeDataAnnotation();` and the key and all other fields with unchanged values still appear.

Our temporary workaround was to do a comparison and removal in the AutoSavePreAction, similar to the below, but this is ugly.

```
AuditManager.DefaultConfiguration.AutoSavePreAction =
    (context, audit) =>
    {
        foreach (var item in audit.Entries)
        {
            var toRemove = item.Properties.Where(p => p.NewValueFormatted == p.OldValueFormatted).ToList();
            foreach (var badItem in toRemove)
                item.Properties.Remove(badItem);

        }
        (context as MyContxt).AuditEntries.AddRange(audit.Entries);
    };
```

Any thoughts?

**Comment by JonathanMagnan** [comment]
Hello @mitchelsellers ,

We just tested it but everything seems fine.

The key is **always** added since it's how you can retrieve from which row values are beyond. However unchanged properties are not added.

Could you try to reproduce it issue a similar code as this one?

```csharp
using System;
using System.Data.SqlClient;
using System.Linq;
using Microsoft.EntityFrameworkCore;
using Z.Test.EntityFramework.Plus;

namespace Z.EntityFramework.Plus.Lab.EFCore20
{
    public static class Request_Audit
    {
        public static void Execute()
        {
            // CLEAR
            using (var ctx = new EntityContext())
            {
                ctx.Entity_Request_Audits.RemoveRange(ctx.Entity_Request_Audits);
                ctx.SaveChanges();
            }

            // SEED
            using (var ctx = new EntityContext())
            {
                ctx.Entity_Request_Audits.Add(new Entity_Request_Audit {ColumnInt = 1, ColumnDate = DateTime.Now});
                ctx.Entity_Request_Audits.Add(new Entity_Request_Audit {ColumnInt = 2, ColumnDate = DateTime.Now});
                ctx.Entity_Request_Audits.Add(new Entity_Request_Audit {ColumnInt = 3, ColumnDate = DateTime.Now});

                ctx.SaveChanges();
            }

            // TEST
            using (var ctx = new EntityContext())
            {
                var audit = new Audit();

                var list = ctx.Entity_Request_Audits.ToList();

                list[0].ColumnInt = -1;
                list[1].ColumnDate = DateTime.Now.AddDays(3);

                ctx.SaveChanges(audit);
            }
        }

        public class EntityContext : DbContext
        {
            static EntityContext()
            {
                AuditManager.DefaultConfiguration.AutoSavePreAction = (context, audit) =>
                    (context as EntityContext).AuditEntries.AddRange(audit.Entries);
            }

            public DbSet<Entity_Request_Audit> Entity_Request_Audits { get; set; }
            public DbSet<AuditEntry> AuditEntries { get; set; }
            public DbSet<AuditEntryProperty> AuditEntryProperties { get; set; }

            protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
            {
                optionsBuilder.UseSqlServer(new SqlConnection(My.Config.ConnectionStrings.TestDatabase));

                base.OnConfiguring(optionsBuilder);
            }
        }

        public class Entity_Request_Audit
        {
            public int ID { get; set; }
            public int ColumnInt { get; set; }

            public DateTime ColumnDate { get; set; }
        }
    }
}
```

**Comment by mitchelsellers** [comment]
The primary key makes sense, I'll adjust our side to make sure that isn't an issue.  THe other columns though are always showing changes, even if there is nothing that was changed.  

I'm testing from an MCS level where we are setting values for everything, as we don't know the changed values vs. the existing values, that might be our issue.  Any workaround?

**Comment by JonathanMagnan** [comment]
Hello @mitchelsellers ,

This depend how you do it.

By example, do you have access to the old value?

Best Regards,

Jonathan

**Comment by mitchelsellers** [comment]
We don't easily have it today.  The audit properly captures the old value for us, it just doesn't differentiate between a changed value and an unchanged value.  

Selectively setting properties for us, only on change, would defeat the purpose of using a utility that helps with Audits.

**Comment by JonathanMagnan** [comment]
I'm not sure to exactly understand what you are saying.

However, if the entity is not taken from the database than no, this library cannot update only modified value since the auditing work with the `ChangeTracker`.

Our paid library `Entity Framework Extensions` do auditing from the database but it's not a rich as this library in term of auditing features.

Anyway, let me know if there is anything in which I can help you.

Best Regards,

Jonathan

**Comment by mitchelsellers** [comment]
Let me try to explain this a little bit better.

We are doing something like this.

```
private void UpdateCustomer(UpdateCustomerViewModel data)
{
var toUpdate = _context.Customers.FirstOrDefault(c => c.CustomerId == data.CustomerId);
if (toUpdate == null)
	return;

toUpdate.CustomerName = data.CustomerName;
toUpdate.Address = data.Address;
toUpdate.City = data.City;
//Etc
_context.Customers.Update(toUpdate);

var audit = new Audit {CreatedBy = editedBy};
_context.SaveChanges(audit);
}
```

In this case CustomerViewModel is a VM for editing, Customer is our entity.  We set all properties, but we only want to audit when the value moves.  

So if the address is set to "123 Main Street" and it is still "123 Main Street" we want to keep it that way.  Make sense?

**Comment by JonathanMagnan** [comment]
Yup make sense so far,

But it should already work like it?

Unless I'm wrong, Entity Framework will not mark this property has modified since the value is equal.

**Comment by mitchelsellers** [comment]
That is not what I see in ASP.NET Core 2.0, with latest version of entity framework.

With the above sample code if I have old values and new values that are exactly the same, I get a log entry showing that it changed, but the values are the same.

Using the code I posted earlier I'm able to suppress it, but that is the only way.

**Comment by JonathanMagnan** [comment]
Do you think you could reproduce it with a similar code as this one?

```csharp
using System;
using System.Data.SqlClient;
using System.Linq;
using Microsoft.EntityFrameworkCore;
using Z.Test.EntityFramework.Plus;

namespace Z.EntityFramework.Plus.Lab.EFCore20
{
    public static class Request_Audit
    {
        public static void Execute()
        {
            // CLEAR
            using (var ctx = new EntityContext())
            {
                ctx.Entity_Request_Audits.RemoveRange(ctx.Entity_Request_Audits);
                ctx.SaveChanges();
            }

            // SEED
            using (var ctx = new EntityContext())
            {
                ctx.Entity_Request_Audits.Add(new Entity_Request_Audit {ColumnInt = 1, ColumnDate = DateTime.Now});
                ctx.Entity_Request_Audits.Add(new Entity_Request_Audit {ColumnInt = 2, ColumnDate = DateTime.Now});
                ctx.Entity_Request_Audits.Add(new Entity_Request_Audit {ColumnInt = 3, ColumnDate = DateTime.Now});

                ctx.SaveChanges();
            }

            // TEST
            using (var ctx = new EntityContext())
            {
                var audit = new Audit();

                var list = ctx.Entity_Request_Audits.ToList();

                list[0].ColumnInt = -1;
                list[0].ColumnDate = list[0].ColumnDate;
                list[1].ColumnInt = list[1].ColumnInt;
                list[1].ColumnDate = DateTime.Now.AddDays(3);

                ctx.SaveChanges(audit);
            }
        }

        public class EntityContext : DbContext
        {
            static EntityContext()
            {
                AuditManager.DefaultConfiguration.AutoSavePreAction = (context, audit) =>
                    (context as EntityContext).AuditEntries.AddRange(audit.Entries);
            }

            public DbSet<Entity_Request_Audit> Entity_Request_Audits { get; set; }
            public DbSet<AuditEntry> AuditEntries { get; set; }
            public DbSet<AuditEntryProperty> AuditEntryProperties { get; set; }

            protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
            {
                optionsBuilder.UseSqlServer(new SqlConnection(My.Config.ConnectionStrings.TestDatabase));

                base.OnConfiguring(optionsBuilder);
            }
        }

        public class Entity_Request_Audit
        {
            public int ID { get; set; }
            public int ColumnInt { get; set; }

            public DateTime ColumnDate { get; set; }
        }
    }
}
```

**Comment by JonathanMagnan** [comment]
Hello @mitchelsellers ,

Do you think you could provide us an example?

Best Regards,

Jonathan

**Comment by JonathanMagnan** [comment]
Hello @mitchelsellers ,

We are currently making a cleanup of some request that we have no answer.

Feel free to reopen if you believe you can help us.

Best Regards,

Jonathan

---
_Help us to keep this library free: [Donate](http://www.zzzprojects.com/contribute)_

**Comment by raublekick** [comment]
I am having this same issue. The way I update entities is nearly exactly how @mitchelsellers does it. His workaround does work out to the desired output.

**Comment by JonathanMagnan** [comment]
Hello @raublekick ,

Do you think you could provide a "working" project example/test with this issue?

It will make easier/faster for my developer to getting started for investigating it.

_You can send it to info@zzzprojects.com if you need to keep the source private_

_We now always ask for a project since we found out that most issues are missing some essential information, have some syntax mistake or are resolved by the requestor when creating it_

_(Even if the issue seem very easy to reproduce, by getting a test project, it allow us to give a faster support and better experience for the support of all our free libraries)_

Best Regards,

Jonathan

---

**Performance Libraries**
`context.BulkInsert(list, options => options.BatchSize = 1000);  `
[Entity Framework Extensions](http://entityframework-extensions.net/) • [Bulk Operations](http://bulk-operations.net/) •  [Dapper Plus](http://dapper-plus.net/) •  [LinqToSql Plus](http://linqtosql-plus.net/)

**Runtime Evaluation**
`Eval.Execute("x + y", new {x = 1, y = 2}); // return 3`
[C# Eval Function](http://eval-expression.net/) • [SQL Eval Function](http://eval-sql.net/)

**Comment by antony7777** [comment]
Hi, I was having the same problem.. All unchanged properties are audited in AuditEntryProperties.
I found that my problem was I called DbContext.Update(model).
Removing it, only keys and changed properties are audited.

This does make sense as according to [documentation](https://docs.microsoft.com/en-us/ef/core/api/microsoft.entityframeworkcore.dbcontext) of DbContext.Update, Update does:
Begins tracking the given entity in the Modified state such that it will be updated in the database when SaveChanges() is called.
**All properties of the entity will be marked as modified.**

I later testes with DbContext.Entity.Update(model), it also audited all unchanged properties.
Hope it's the root cause.
